src/librustc/dep_graph/graph.rs

```diff
@@ -518,29 +565,59 @@ impl DepGraph {
     }
 
     pub fn serialize(&self) -> SerializedDepGraph {
-        let current = &self.data.as_ref().unwrap().current;
+        let data = self.data.as_ref().unwrap();
+
+        let current = &data.current;
         let shards = current.data.lock_shards();
-        let node_count = current.node_count.load(Acquire) as usize;
-        let data = || (0..node_count).map(|i| {
+        let node_count = data.node_count();
+        let node_data_iter = || (0..node_count).map(|i| {
             let shard = i % sharded::SHARDS;
             let inner = i / sharded::SHARDS;
             &shards[shard][inner]
         });
 
-        let fingerprints: IndexVec<SerializedDepNodeIndex, _> =
-            data().map(|d| d.fingerprint).collect();
-        let nodes: IndexVec<SerializedDepNodeIndex, _> =
-            data().map(|d| d.node).collect();
+        let fingerprints: IndexVec<DepNodeIndex, _> =
+            node_data_iter().map(|d| d.fingerprint).collect();
+        let mut nodes: IndexVec<DepNodeIndex, _> =
+            node_data_iter().map(|d| d.node).collect();
+
+        // Invalidate dep nodes with unknown state as these cannot safely
+        // be marked green in the next session. One of the dependencies of the
+        // unknown node may have changed in this session (and is currently marked red),
+        // but might be green again in the next session, which may cause the unknown node
+        // to incorrectly be marked green in the next session, even though one of its dependencies
+        // did actually change.
```

Comment:
OK, the logic in this comment lost me. I think I get the outlines, but there's some subtle stuff going on here -- is there a summary somewhere of how this is meant to work?